lcls-twincat-common-components/Library/Devices/XPIM/FB_XPIM.TcPOU

```diff
@@ -0,0 +1,231 @@
+    bEnableMotion: BOOL;
+    // Set this to TRUE to enable beam parameter checks, or FALSE to disable them.
+    bEnableBeamParams: BOOL;
+    // Set this to TRUE to enable position limit checks, or FALSE to disable them.
+    bEnablePositionLimits: BOOL;
+    // The name of the device for use in the PMPS DB lookup and diagnostic screens.
+    sDeviceName: STRING;
+    // The name of the transition state in the PMPS database.
+    sTransitionKey: STRING;
+    // Set this to TRUE to re-read the loaded database immediately (useful for debug).
+    bReadDBNow: BOOL;
+
+    // While TRUE, the zoom motor cannot be moved.
+    {attribute 'pytmc' := '
+        pv: CLZ:LOCK
+        io: io
+        field: ZNAM Unlocked
+        field: ONAM Locked
+    '}
+    bZoomLock: BOOL;
+
+    // While TRUE, the focus motor cannot be moved.
+    {attribute 'pytmc' := '
+        pv: CLF:LOCK
+        io: io
+        field: ZNAM Unlocked
+        field: ONAM Locked
+    '}
+    bFocusLock: BOOL;
+
+    // Forward limit disable for zoom
+    bZoomEndFwd AT %I*: BOOL;
+    // Backward limit disable for zoom
+    bZoomEndBwd AT %I*: BOOL;
+    // Forward limit disable for focus
+    bFocusEndFwd AT %I*: BOOL;
+    // Backward limit disable for focus
+    bFocusEndBwd AT %I*: BOOL;
+END_VAR
+VAR_OUTPUT
+    // The current position state as an enum.
+    {attribute 'pytmc' := '
+        pv: MMS:STATE:GET
+        io: i
+    '}
+    eEnumGet: E_XPIM_States;
+    // The PMPS database lookup associated with the current position state.
+    stDbStateParams: ST_DbStateParams;
+END_VAR
+VAR
+    bInit: BOOL;
+
+    fbYStage: FB_MotionStage;
+    fbZoom: FB_MotionStage;
+    fbFocus: FB_MotionStage;
+
+    fbStateDefaults: FB_PositionState_Defaults;
+
+    {attribute 'pytmc' := '
+        pv: MMS
+        astPositionState.array: 1..4
+    '}
+    fbStates: FB_PositionStatePMPS1D;
+    astPositionState: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+    fbArrCheckWrite: FB_CheckPositionStateWrite;
+
+    {attribute 'pytmc' := 'pv: MFW'}
+    fbFilterWheel: FB_XPIM_FilterWheel;
+
+    {attribute 'pytmc' := 'pv: CAM'}
+    fbOpal: FB_XPIM_Opal;
+
+    {attribute 'pytmc' := 'pv: CIL'}
+    fbLED: FB_XPIM_LED;
+
+    {attribute 'pytmc' := 'pv: SFW'}
+    fbFlowSwitch: FB_XTES_Flowswitch;
+END_VAR
+VAR CONSTANT
+    // State defaults if not provided
+    fDelta: LREAL := 2;
+    fAccel: LREAL := 200;
+    fOutDecel: LREAL := 25;
+END_VAR]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[
+IF NOT bInit THEN
+    bInit := TRUE;
+
+    stYStage.bHardwareEnable := TRUE;
+    stYStage.bPowerSelf := FALSE;
+    // No limit switch at the bottom
+    stYStage.bLimitBackwardEnable := TRUE;
+
+    // Extra lock on lens + lens limits are NO instead of NC
+    stZoomStage.bHardwareEnable := NOT bZoomLock;
+    stZoomStage.bPowerSelf := TRUE;
+    stZoomStage.bLimitForwardEnable := NOT bZoomEndFwd;
```

Comment:
These limit settings only happen at init time?

Reply:
This is a big error, good catch